### ISSUE TITLE
WPF: Plot does not update when inside Material Design Card

### ISSUE THREAD

**Issue by swharden** [ScottPlot/ScottPlot#1733]
This was emailed-in from Ryan and copied here (with permission):

I'm not a WPF export and haven't used Material Design before, so perhaps someone with more experience would be better at addressing this issue.

If I don't hear anything after a few days I'll give it a go and see if I can figure it out.

---

I'm writing this email to inform you that there's an issue with the plots when trying to .Render() or .Refresh() when the WPF plot is within a [Material Design XAML](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit) Card.

The plot's initial render works, but any modification to the plot, and calls to render or refresh are basically ignored, and the plot does not update.

Breakpointing through the code I use to update the chart, everything works fine, the plottables are filled, and render/refresh is called without any exceptions.

Is there a debug mode I can enable to see more information maybe?

To quickly reproduce this bug, I have a WPFPlot in a usercontrol like so:

```xml
<UserControl>
    x:Class="Program.Plots.DataPlot"
    xmlns="http://[schemas.microsoft.com](http://schemas.microsoft.com/)/winfx/2006/xaml/presentation"
    xmlns:x="http://[schemas.microsoft.com](http://schemas.microsoft.com/)/winfx/2006/xaml"
    xmlns:d="http://[schemas.microsoft.com](http://schemas.microsoft.com/)/expression/blend/2008"
    xmlns:local="clr-namespace:Program.Plots"
    xmlns:mc="http://[schemas.openxmlformats.org](http://schemas.openxmlformats.org/)/markup-compatibility/2006"
    d:DesignHeight="450"
    d:DesignWidth="800"
    mc:Ignorable="d">
    <Grid MouseDown="Grid_MouseDown">
        <WpfPlot Name="WpfPlot1" />
    </Grid>
</UserControl>
```

And in the code behind, I have this code: 

```cs
public partial class DataPlot : UserControl
    {
        public DataPlot()
        {
            InitializeComponent();
           
            for (int i = 0; i < WpfPlot1.Plot.Palette.Count(); i++)
            {
                double[] xs = DataGen.Consecutive(100);
                double[] ys = DataGen.Sin(100, phase: -i * .5 / WpfPlot1.Plot.Palette.Count());
                WpfPlot1.Plot.AddScatterLines(xs, ys, lineWidth: 3);
            }

            WpfPlot1.Plot.Title($"{WpfPlot1.Plot.Palette}");
            WpfPlot1.Plot.AxisAuto(0, 0.1);
            WpfPlot1.Plot.Style(ScottPlot.Style.Gray2);
            var bnColor = System.Drawing.ColorTranslator.FromHtml("#2e3440");
            WpfPlot1.Plot.Style(figureBackground: bnColor, dataBackground: bnColor);

            WpfPlot1.Refresh();
        }

        private void Grid_MouseDown(object sender, MouseButtonEventArgs e)
        {
           
            WpfPlot1.Plot.Clear();
            WpfPlot1.Refresh();
        }
   } 
```

When you click on the grid, mouse down, it clears, only if the UserControl is outside of a Material Design Card. If it is inside the card, even through the clear function and render function is ran, the visual of the chart never clears/changes.

### This works

```xml
<plots:DataPlot></plots:DataPlot>
<materialDesign:Card
    Name="MainMenuCard"
    Grid.Row="1"
    Margin="16"
    Padding="24">
</materialDesign:Card>
```

### This fails

```xml
<materialDesign:Card
    Grid.Row="1"
    Margin="16"
    Padding="24">
    <plots:DataPlot></plots:DataPlot>
</materialDesign:Card>
```

The above isn't my end-game implementation of ScottPlot, it's just a way to quickly reproduce the bug. My actual implementation involves instantiating a WPFPlot in a Viewmodel, and setting the content of a <ContentPresenter> to the WPFPlot property of the viewmodel. The result is the same, it renders fine the first time, but doesn't render anything new after that. My program is mostly Cards so since it exists in just one card, it doesn't work.

**Comment by swharden** [comment]
When `WpfPlot.Refresh()` is called, here's the relevant code that is called:

https://github.com/ScottPlot/ScottPlot/blob/3581f8c5aaad0b0c46c4dcc87e15013ea71abb59/src/ScottPlot4/ScottPlot.WPF/WpfPlot.xaml.cs#L142-L150

https://github.com/ScottPlot/ScottPlot/blob/3581f8c5aaad0b0c46c4dcc87e15013ea71abb59/src/ScottPlot4/ScottPlot/Control/Backend.cs#L313-L365

**Comment by swharden** [comment]
Related: #1731

Similar behavior was reported by the person who originally e-mailed me. They've been made aware of this issue thread and have begun trying to diagnose the issue. I'll follow-up here with findings if/when I hear back 👍

**Comment by ChrisCC6** [comment]
Hi @swharden,

I got a test computer yesterday, showing the behavior I wrote in #1731. Updating the graphic driver fixed it. Maybe Ryan's issue has a similar root cause.

Best Regards,
Chris

**Comment by swharden** [comment]
Thanks for following-up @ChrisCC6! That's such a surprising behavior/cause, I might never would have figured it out 😅 Thanks for sharing your findings here!

I reached back to Ryan (who recently emailed me indicating no progress) to make them aware of that potential solution.

Since this ended-up about as far as we can advance it for now, I'll close this issue now 👍

**Comment by swharden** [comment]
I received this email Tue, Apr 12, 3:49 PM

---

Hey Scott, writing this email to update you:

I updated to the most recent Intel Graphics Drivers, and still no dice.

I tried out my application on my Desktop , which is a CPU 5950x/ GPU 3090 build, and it still doesn't update when rendering a new graph. But what is different, is that I'm able to actually use my mouse to move the graph on my desktop, causing it to render the new graph after moving the graph around.

On my dev laptop, with intel integrated graphics, and a touchpad, I can't even use my mouse to move the graph around, and this happens in the example cookbook application as well. 
The part that is weird, is that in the cookbook, there's a mouse tracking example, which clearly shows where my mouse is in relation to being on the graph, but left MB down and holding and dragging cannot move the graph.

In this video, I launch some new data on the graph, and then use the WpfPlotViewer to show a new window with the new plot data, which has the new updated graph on it. But the main graph doesn't change until I use my mouse to change it, this is on my desktop machine with a Nvidia 3090 GPU. On my laptop, I can't even use the mouse to move any plots.

![image](https://user-images.githubusercontent.com/4165489/166162072-55508725-d3a7-4678-ab64-b471ca6c6c3b.png)

The PlotControl doesn't update, and a window shows with the new correct data. 
Moving the plot with my mouse causes it to update. 
Resizing the window causes it to update.
Cannot move plot at all using a laptop touchpad.

https://streamable.com/7xnk8k

**Comment by swharden** [comment]
I received this email Apr 14, 2022, 3:35 PM

---

Hey Scott,

I know you're probably tired of hearing from me, but I found a way to get around this issue.

I just new the plot object when I get new data.

Whenever I have new data come in (Not real time of course because that would be horrendous), I just fire up a dispatcher thread, and do the following: 

![image](https://user-images.githubusercontent.com/4165489/166162110-27ac1412-a0b6-40e3-8e51-581139712d75.png)

Within that foreach, I add my data to the plot using this.PlotControl.Plot.AddSignal,

Then call :

```cs
this.PlotControl.Refresh();
this.PlotControl.Plot.AxisAuto();
this.PlotControl.Plot.Render();
```

And since I have the plot rendering through a <ContentControl> that is set to update on UpdateSourceTrigger=PropertyChanged, and this.PlotControl is registered as a property that fires OnPropertyChanged, the graph renders as desired.

I've tried setting the program to software rendering only and that didn't work.
I tried just calling OnPropertyChanged for the PlotControl, and that didn't work either.

The only thing that has worked, is just newing the object entirely, resetting it up, refilling it, and refreshing and rendering.

The plot still doesn't work with a touchpad on any laptop, but that's fine, not that important. Plugging in a mouse works.
